src/v/kafka/server/group.h

```diff
@@ -734,6 +736,10 @@ class group final : public ss::enable_lw_shared_from_this<group> {
       model::topic_partition,
       std::unique_ptr<offset_metadata_with_probe>>
       _offsets;
+    group_probe<
+      model::topic_partition,
+      std::unique_ptr<offset_metadata_with_probe>>
+      _probe;
```

Comment:
@NyaliaLui @VladLazar

A group is generally managed as a shared pointer:

`class group final : public ss::enable_lw_shared_from_this<group> {`

because there is some asynchronous memory management. for example, a group might be deleted, but it sticks aroudn because maybe there are background operations that are hanging on to a reference to the group.

this is related to metrics because it seems like there might be a metrics registration/de-registration race. deregistration happens when the group destructor runs which might happen after a group that has the same identity in the metrics registration is created. i think that seastar in this case will assert-fail.

should be something that a test can be written for to demonstrate the issue. wdyt?

Reply:
I'll restate the scenario just to make sure i got it right:
* a group is created and stored somewhere
* shared_ptr references to that group are handed out
* the group is deleted from wherever it was stored
* the outstanding references extend the lifetime of the group
* a new identical group is created

I'd expect to see a `double_registration` exception thrown in the last step. A `metric_groups` object doesn't have an identity; it's just a container of metadata for the metrics registered with it. Also, there doesn't seem to be any assertion in the seastar metrics layer, so maybe I'm misunderstanding your point.

Reply:
I don't really understand how this test should look like. Do we wish to test that a double registration exception is thrown in the scenario above, or did you have something else in mind?

Reply:
@VladLazar it's not an assertion like i alluded to, but rather an exception. we have lots of noexcept interfaces so throws often are hard crashes. here is the double reigstration scenario:

https://github.com/scylladb/seastar/blob/master/src/core/metrics.cc#L355-L357

Reply:
Ah right. I see the mapping between the exception and assertion now. Sure, we can add a test, but fundamentally, if we should be able to have two virtually identically groups alive at any given time, then maybe this is not the right place for this probe.

Reply:
it might still be the right place, and we just need to separate registeration from lifetime issues. in this case the issue is mostly that deregistration is tied to a destructor. so maybe we need to make sure that metrics.clear() or whatever is called at an appropriate time? im not sure, but its something that won't be good to hit in production if it is possible.

Reply:
I don't see how it would be possible to separate deregistration from  the lifetime. The metrics registered by a `group`, can't be cleared from another instance.

Reply:
I had a look at how the `group` type is used. Instances are stored in the `group_manager` and handles are not handed out outside of the class. A double registration can occur if an existing group is deleted while an operation is being performed on it (e.g. transaction, sync, etc.) and a new identical group (same group id, for the same topic and partition) is created before the ongoing operation that's keeping the old group alive finishes.

@dotnwat is that a possible scenario? We could guard against it by clearing the probe upon erasing a group from the `group_manager`.